api/transport/stream.go

```diff
@@ -0,0 +1,165 @@
+}
+
+// Request contains all the metadata about the request.
+func (s *ServerStream) Request() *StreamRequest {
+	return s.stream.Request()
+}
+
+// SendMessage sends a request over the stream. It blocks until the message
+// has been sent.  In certain implementations, the timeout on the context
+// will be used to timeout the request.
+func (s *ServerStream) SendMessage(ctx context.Context, msg *StreamMessage) error {
+	return s.stream.SendMessage(ctx, msg)
+}
+
+// ReceiveMessage blocks until a message is received from the connection. It
+// returns an io.Reader with the contents of the message.
+func (s *ServerStream) ReceiveMessage(ctx context.Context) (*StreamMessage, error) {
+	return s.stream.ReceiveMessage(ctx)
+}
+
+type clientStreamOpts struct{}
+
+// ClientStreamOption is an option for configuring a client stream.
+type ClientStreamOption interface {
+	apply(*clientStreamOpts)
+}
+
+// NewClientStream will create a new ClientStream.
+func NewClientStream(s StreamWithClose, options ...ClientStreamOption) (*ClientStream, error) {
+	if s == nil {
+		return nil, yarpcerrors.InvalidArgumentErrorf("non-nil stream with close is required")
+	}
+	return &ClientStream{stream: s}, nil
+}
+
+// ClientStream represents the Client API of interacting with a Stream.
+type ClientStream struct {
+	stream StreamWithClose
+}
+
+// Context returns the context for the stream.
+func (s *ClientStream) Context() context.Context {
+	return s.stream.Context()
+}
+
+// Request contains all the metadata about the request.
+func (s *ClientStream) Request() *StreamRequest {
+	return s.stream.Request()
+}
+
+// SendMessage sends a request over the stream. It blocks until the message
+// has been sent.  In certain implementations, the timeout on the context
+// will be used to timeout the request.
+func (s *ClientStream) SendMessage(ctx context.Context, msg *StreamMessage) error {
+	return s.stream.SendMessage(ctx, msg)
+}
+
+// ReceiveMessage blocks until a message is received from the connection. It
+// returns an io.Reader with the contents of the message.
+func (s *ClientStream) ReceiveMessage(ctx context.Context) (*StreamMessage, error) {
+	return s.stream.ReceiveMessage(ctx)
+}
+
+// Close will close the connection. It blocks until the server has
+// acknowledged the close. In certain implementations, the timeout on the
+// context will be used to timeout the request. If the server timed out the
+// connection will be forced closed by the client.
+func (s *ClientStream) Close(ctx context.Context) error {
+	return s.stream.Close(ctx)
+}
+
+// StreamWithClose represents an API of interacting with a Stream that is
+// closable.
+type StreamWithClose interface {
+	Stream
+
+	// Close will close the connection. It blocks until the server has
+	// acknowledged the close. In certain implementations, the timeout on the
+	// context will be used to timeout the request. If the server timed out the
+	// connection will be forced closed by the client.
```

Comment:
nit: instead of saying "in certain implementations" here and below, you can just say, "The provided context controls the timeout for this operation if the implementation supports it."